File: class19/python/TCC1901-ShiftReduce.py
#!/usr/bin/env python3
"""TCC1901-ShiftReduce.py | Aula 19
Simulador de análise sintática shift-reduce (bottom-up).
Gramática:
  E → E + E
  E → E * E
  E → ( E )
  E → id
"""
from typing import List, Tuple


# Produções (lado esquerdo, lado direito)
producoes = [
    ("E", ["E", "+", "E"]),
    ("E", ["E", "*", "E"]),
    ("E", ["(", "E", ")"]),
    ("E", ["id"]),
]


def reduzir(pilha: List[str]) -> bool:
    """Tenta reduzir o topo da pilha. Retorna True se reduziu."""
    for esq, dir_ in producoes:
        if len(pilha) >= len(dir_) and pilha[-len(dir_):] == dir_:
            del pilha[-len(dir_):]
            pilha.append(esq)
            return True
    return False


def shift_reduce(entrada: List[str]) -> Tuple[bool, List[str]]:
    """Executa análise shift-reduce. Retorna (aceita, trace)."""
    pilha: List[str] = []
    entrada = entrada + ["$"]
    i = 0
    trace: List[str] = []
    trace.append(f"Início: pilha={pilha}, entrada={entrada[i:]}")

    while True:
        # Tenta reduzir o máximo possível
        while reduzir(pilha):
            trace.append(f"Reduce: pilha={pilha}, entrada={entrada[i:]}")

        # Verifica aceitação: pilha = [E], entrada = [$]
        if pilha == ["E"] and entrada[i] == "$":
            trace.append("ACEITA")
            return True, trace

        # Shift
        if entrada[i] != "$":
            pilha.append(entrada[i])
            i += 1
            trace.append(f"Shift:  pilha={pilha}, entrada={entrada[i:]}")
        else:
            trace.append("REJEITA (entrada esgotada sem aceitação)")
            return False, trace


def tokenize(expr: str) -> List[str]:
    tokens = []
    i = 0
    while i < len(expr):
        if expr[i].isspace():
            i += 1
            continue
        if expr[i].isdigit():
            tokens.append("id")
            while i < len(expr) and expr[i].isdigit():
                i += 1
        elif expr[i] in "+*()":
            tokens.append(expr[i])
            i += 1
        else:
            raise SyntaxError(f"Caractere inválido: {expr[i]!r}")
    return tokens


if __name__ == "__main__":
    testes = ["id + id", "id * id + id", "( id + id ) * id", "id +"]
    for expr in testes:
        tokens = tokenize(expr)
        ok, trace = shift_reduce(tokens)
        print(f"{'ACEITA' if ok else 'REJEITA'}: {expr}")
        if expr == "id + id":
            for linha in trace:
                print(f"  {linha}")

File: class19/java/TCC1901-ShiftReduce.java
// TCC1901-ShiftReduce.java | Aula 19
// Simulador de análise sintática shift-reduce (bottom-up).
// Espelho Java do TCC1901-ShiftReduce.py (mesmas entradas e saídas).
import java.util.*;

public class TCC1901ShiftReduce {

    // Produções (lado esquerdo, lado direito)
    static String[][] producoes = {
        {"E", "E", "+", "E"},
        {"E", "E", "*", "E"},
        {"E", "(", "E", ")"},
        {"E", "id"},
    };

    static boolean reduzir(List<String> pilha) {
        for (String[] prod : producoes) {
            String esq = prod[0];
            int len = prod.length - 1;
            if (pilha.size() >= len) {
                boolean match = true;
                for (int k = 0; k < len; k++) {
                    if (!pilha.get(pilha.size() - len + k).equals(prod[k + 1])) {
                        match = false;
                        break;
                    }
                }
                if (match) {
                    for (int k = 0; k < len; k++) pilha.remove(pilha.size() - 1);
                    pilha.add(esq);
                    return true;
                }
            }
        }
        return false;
    }

    static class Resultado {
        boolean aceita;
        List<String> trace;
        Resultado(boolean a, List<String> t) { aceita = a; trace = t; }
    }

    static Resultado shiftReduce(List<String> entrada) {
        List<String> pilha = new ArrayList<>();
        List<String> ent = new ArrayList<>(entrada);
        ent.add("$");
        int i = 0;
        List<String> trace = new ArrayList<>();
        trace.add("Início: pilha=" + pilha + ", entrada=" + ent.subList(i, ent.size()));

        while (true) {
            while (reduzir(pilha)) {
                trace.add("Reduce: pilha=" + pilha + ", entrada=" + ent.subList(i, ent.size()));
            }
            if (pilha.size() == 1 && pilha.get(0).equals("E") && ent.get(i).equals("$")) {
                trace.add("ACEITA");
                return new Resultado(true, trace);
            }
            if (!ent.get(i).equals("$")) {
                pilha.add(ent.get(i));
                i++;
                trace.add("Shift:  pilha=" + pilha + ", entrada=" + ent.subList(i, ent.size()));
            } else {
                trace.add("REJEITA (entrada esgotada sem aceitação)");
                return new Resultado(false, trace);
            }
        }
    }

    static List<String> tokenize(String expr) {
        List<String> tokens = new ArrayList<>();
        int i = 0, n = expr.length();
        while (i < n) {
            char c = expr.charAt(i);
            if (Character.isWhitespace(c)) { i++; continue; }
            if (Character.isDigit(c)) {
                tokens.add("id");
                while (i < n && Character.isDigit(expr.charAt(i))) i++;
            } else if ("+*()".indexOf(c) >= 0) {
                tokens.add(String.valueOf(c));
                i++;
            } else {
                throw new RuntimeException("Caractere inválido: '" + c + "'");
            }
        }
        return tokens;
    }

    public static void main(String[] args) {
        String[] testes = {"id + id", "id * id + id", "( id + id ) * id", "id +"};
        for (String expr : testes) {
            List<String> tokens = tokenize(expr);
            Resultado r = shiftReduce(tokens);
            System.out.println((r.aceita ? "ACEITA" : "REJEITA") + ": " + expr);
            if (expr.equals("id + id")) {
                for (String linha : r.trace) {
                    System.out.println("  " + linha);
                }
            }
        }
    }
}
